homeassistant/components/cover/__init__.py

```diff
@@ -260,17 +269,21 @@ async def async_close_cover(self, **kwargs):
 
     def toggle(self, **kwargs: Any) -> None:
         """Toggle the entity."""
-        if self.is_closed:
+        if SUPPORT_STOP | self.supported_features and (
+            self.is_closing or self.is_opening
+        ):
+            self.stop_cover(**kwargs)
+        elif self.is_closed:
             self.open_cover(**kwargs)
         else:
-            self.close_cover(**kwargs)
+            if self.is_last_toggle_direction_open:
+                self.close_cover(**kwargs)
+            else:
+                self.open_cover(**kwargs)
 
     async def async_toggle(self, **kwargs):
         """Toggle the entity."""
-        if self.is_closed:
-            await self.async_open_cover(**kwargs)
-        else:
-            await self.async_close_cover(**kwargs)
+        await self.hass.async_add_executor_job(ft.partial(self.toggle, **kwargs))
```

Comment:
This will add an unnecessary deferment of the toggle for integrations that are async friendly (and some tests are failing because of that).

Add a helper function to avoid duplicating the logic, maybe something like:
```py
def _get_toggle_function(self, fns):
    if SUPPORT_STOP | self.supported_features and (
        self.is_closing or self.is_opening
    ):
        return fns["stop"]
    elif self.is_closed:
        return fns["open"]
    else:
        if self.is_last_toggle_direction_open:
        return fns["close"]
        else:
        return fns["open"]
```

And call it like this:
```py
    async def async_toggle(self, **kwargs):
        """Toggle the entity."""
        fns = {"open": self.open_cover, "close": self.close_cover, "stop": self.stop_cover}
        fn = self._get_toggle_function(fns)
        await self.hass.async_add_executor_job(ft.partial(fn, **kwargs))
```

Reply:
This will work very well, thank you for the detailed suggestion!